### PR TITLE
Android:Make MessageQueueThreadImpl startNewBackgroundThread more eff…

### PR DESCRIPTION
**Pull request by chnfeeeeeef**
**Notes**
By adding thread synchronization,creating background threads is more efficient.
Thanks for @javache 

TianXia
XiaoMi Corp.

### CODE REVIEW

**Comment by javache** [comment]
Thanks for the PR! Can you explain why this is more efficient? I can also import this and run some internal perf tests on this.

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java [hunk 1]
```diff
@@ -184,27 +184,8 @@ private static MessageQueueThreadImpl startNewBackgroundThread(
     final String name,
     long stackSize,
     QueueThreadExceptionHandler exceptionHandler) {
-    final SimpleSettableFuture<Looper> looperFuture = new SimpleSettableFuture<>();
-    final SimpleSettableFuture<MessageQueueThread> mqtFuture = new SimpleSettableFuture<>();
-    Thread bgThread = new Thread(null,
-        new Runnable() {
-          @Override
-          public void run() {
-            Process.setThreadPriority(Process.THREAD_PRIORITY_DISPLAY);
```
Comment (javache) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java]: Where did this go?

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java [hunk 2]
```diff
@@ -184,27 +184,8 @@ private static MessageQueueThreadImpl startNewBackgroundThread(
...
-            Looper.prepare();
-
-            looperFuture.set(Looper.myLooper());
-            MessageQueueThreadRegistry.register(mqtFuture.getOrThrow());
```
Comment (javache) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java]: Where did this go?

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  @Override
+  public void run() {
+    Looper.prepare();
+    synchronized (this) {
+      mLooper = Looper.myLooper();
+      notifyAll();
+    }
+    onLooperPrepared();
```
Comment (javache) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java]: Unused?

Comment (chnfeeeeeef) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java]: [Thread is initialized by the Runnable parameter in the previous code to ensure that Looke is ready when MessageQueueThreadImpl is created]
We can use the thread lock way to solve this problem. MessageQueueThreadImpl initialization, call MessageQueueHandlerThread class getLooper method, getLooper method will block the current thread until the MessageQueueHandlerThread thread is created, so more timely and effective

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java (hunk 2 above)
Comment (chnfeeeeeef) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java]: `return new MessageQueueThreadImpl(name, messageQueueHandlerThread.getLooper(), exceptionHandler);`
The above code can replace the deleted code to ensure that a return with a secure Looper MessageQueueThreadImpl

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java (hunk 1 above)
Comment (chnfeeeeeef) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java]: `Process.setThreadPriority(Process.THREAD_PRIORITY_DISPLAY);`
For a background thread, I feel it is not necessary to improve thread priority

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java (hunk 3 above)
Comment (javache) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java]: Yes, I see that. I'm asking specifically here about the `onLooperPrepared` method.

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java (hunk 2 above)
Comment (javache) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java]: Well, we should do another change first then to clean that up to make sure that's true.

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java (hunk 1 above)
Comment (javache) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java]: It's actually quite important to prioritize bridge startup. I believe we have some changes that reduce the priority after startup has completed.

Comment (chnfeeeeeef) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadImpl.java]:
```
final SimpleSettableFuture<Looper> looperFuture = new SimpleSettableFuture<>();
final SimpleSettableFuture<MessageQueueThread> mqtFuture = new SimpleSettableFuture<>();
```
Doing so for additional performance overhead.

File: ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java (hunk 3 above)
Comment (chnfeeeeeef) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java]: Yes,onLooperPrepared is Unused.

Comment (chnfeeeeeef) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java]: Should I remove` onLooperPrepared ` method ,add `Process.setThreadPriority(Process.THREAD_PRIORITY_DISPLAY)`code  to `MessageQueueHandlerThread ` and pull request again?

Comment (javache) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java]: Yes, please do so.

I'll look into removing MessageQueueThreadRegistry

Comment (chnfeeeeeef) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueHandlerThread.java]: I have removed unused onLooperPrepared method,added Process.setThreadPriority(Process.THREAD_PRIORITY_DISPLAY) code into MessageQueueHandlerThread and updated the code.

**Comment by javache** [comment]
Can  you please sign the CLA so I can import this?

**Comment by chnfeeeeeef** [comment]
Looking forward to import. :) 
Thank you for your reviews. @javache

**Comment by javache** [comment]
I've run some internal tests on this and there's no obvious performance win here. Do you have some numbers from benchmarks yourself on this?

**Comment by chnfeeeeeef** [comment]
I've also run some internal tests after removing the code` MessageQueueThreadRegistry.register(mqtFuture.getOrThrow())`,seems no obvious performance win too.
To ensure that a return with a secure Looper for MessageQueueThreadImpl,I think `MessageQueueHandlerThread` is more clear.

**Pull request closed without merging**